metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -264,7 +264,31 @@ void sum_and_call_counts(const fs::path &dir,
     }
 }
 
-rd_succ_bv_type route_at_forks(const graph::DBGSuccinct &graph,
+uint64_t from_graph_index(const graph::DeBruijnGraph &graph,
+                          graph::DeBruijnGraph::node_index idx) {
+    if (auto* g = dynamic_cast<graph::DBGSuccinct const*>(&graph)) {
+        return g->kmer_to_boss_index(idx);
+    } else {
+        return idx;
+    }
+}
+graph::DeBruijnGraph::node_index to_graph_index(const graph::DeBruijnGraph &graph,
+                                                uint64_t idx) {
+    if (auto* g = dynamic_cast<graph::DBGSuccinct const*>(&graph)) {
+        return g->boss_to_kmer_index(idx);
+    } else {
+        return idx;
+    }
+}
+size_t get_last_size(const graph::DeBruijnGraph &graph) {
+    if (dynamic_cast<graph::DBGSuccinct const*>(&graph)) {
+        return graph.get_last()->size();
+    } else {
+        return graph.max_index() + 1;
+    }
+}
```

Comment:
I feel like it's idiomatically better to make these virtual functions, but I'm not sure if it's appropriate to place them in `graph::DeBruijnGraph`?

Reply:
`get_last()` is only used in the row diff builder and DBGSuccinct's implementation, so I don't think that warrants making it part of `DeBruijnGraph`.

One way around all of this would be to change the functionality of the dummy mask so that it keeps the indices the same, but it's simply used to avoid traversing to dummy edges and to keep a count of the number of real nodes.

Reply:
It'll mean that node indices in DBGSuccinct are not contiguous, but I think that's the lesser of two evils